### PR TITLE
418 deceased patient

### LINKED ISSUE #418: Add deaths fields to patient schema
- [x] Add isDeceased, dateOfDead fields to patient json schema and update json ui
- [x] Store isDeceased flag in names table
- [x] Show deceased flag in patient list view

### PR DESCRIPTION
**Pull request by clemens-msupply**
- I show the deceased flag in the patient list view right now. Eventually this can be removed or replace by a general patient status flag but though it good to show which information we have at this stage.
- JSONForms controls are not shown if the hidden flag is set (needed for isDeceased json schema rule)

Note, json schema / ajv (not JSONForms) ignores rules based on undefined values. For this reason the `isDeceased: boolean` is required, i.e. it needs to be set to make the ui schema rule work.
https://github.com/ajv-validator/ajv/issues/913

Closes #418

### CODE REVIEW

**Comment by mark-prins** [comment]
Tripped up again.. database migrations would be super helpful.
( I have three databases on the go, and don't recreate every time I switch branches, because that would require getting 4D running and opening the right datafile etc.. )

image

```
{
    "data": null,
    "errors": [
        {
            "message": "\"UNKNOWN\" (\"\\\"\\\\\\\"column name.is_deceased does not exist\\\\\\\"\\\"\")",
            "locations": [
                {
                    "line": 25,
                    "column": 3
                }
            ],
            "path": [
                "me",
                "stores",
                "nodes",
                0
            ]
        }
    ]
}
```

**Comment by mark-prins** [comment]
Or maybe the f/e should be parsing the error better 🤔

File: server/service/src/document/patient/patient_updated.rs [hunk 1]
```diff
@@ -163,6 +164,8 @@ mod test {
                 occupation: None,
             },
             allergies: None,
+            is_deceased: false,
+            date_of_dead: None,
```
Comment (mark-prins) [inline server/service/src/document/patient/patient_updated.rs:168]: :nit 
```suggestion
            date_of_death: None,
```

File: server/service/src/document/patient/update.rs [hunk 2]
```diff
@@ -222,6 +222,8 @@ pub mod test {
                 occupation: None,
             },
             allergies: None,
+            is_deceased: false,
+            date_of_dead: None,
```
Comment (mark-prins) [inline server/service/src/document/patient/update.rs:226]: :nit - as noted
```suggestion
            date_of_death: None,
```

File: server/service/src/document/schemas/patient.json [hunk 3]
```diff
@@ -177,6 +177,11 @@
           "description": "Date of birth is estimated",
           "type": "boolean"
         },
+        "dateOfDead": {
```
Comment (mark-prins) [inline server/service/src/document/schemas/patient.json:180]:
```suggestion
        "dateOfDeath": {
```

File: server/service/src/document/schemas/patient.json [hunk 4]
```diff
@@ -235,6 +244,11 @@
...
```
Comment (mark-prins) [inline server/service/src/document/schemas/patient.json:247]:
```suggestion
        "dateOfDeath": {
```

File: server/service/src/sync/init_programs_data.rs [hunk 5]
```diff
@@ -52,6 +52,8 @@ fn person_1() -> Person {
             literate: None,
             occupation: None,
         },
+        is_deceased: false,
+        date_of_dead: None,
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:56]:
```suggestion
        date_of_death: None,
```

File: server/service/src/sync/init_programs_data.rs [hunk 6]
```diff
@@ -92,6 +96,8 @@ fn person_3() -> Person {
...
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:100]:
```suggestion
        date_of_death: None,
```

File: server/service/src/sync/init_programs_data.rs [hunk 7]
```diff
@@ -72,6 +74,8 @@ fn person_2() -> Person {
...
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:78]:
```suggestion
        date_of_death: None,
```

**Review by mark-prins — APPROVED** [review]
👍  Nice! that works well.

My only suggestion is another one of those annoying spelling / grammar observations, which I got a little carried away with. Haven't noted them all.. but I'm sure you get the idea.

**Comment by clemens-msupply** [comment]
> Or maybe the f/e should be parsing the error better thinking

I am not sure if a migration would fully solve it, e.g. if the DB has an additional column which causes problem in another branch, the migration wouldn't automatically fix it... A message on the FE that there was a DB error would be good though because I also debugged this problem too many times :)

**Comment by clemens-msupply** [comment]
Thanks I fixed the spelling

**Comment by mark-prins** [comment]
The problem will be introduced in a specific branch though - maybe in that one we could migrate? 
I think of it especially when I write scripts to do migrations, run them and then throw them away.

e.g.
```
alter table "name" add column if not exists is_deceased bool null;
update "name" set is_deceased = false;
alter table "name" alter column is_deceased set not null;
```
And still have the table creation with the full version, as it does now.
It would mean messy migration scripts, with lots of unneeded code, so yes. I'd like to avoid that too.
Just expressing frustration.. I've been caught many times with database changes.

**Comment by clemens-msupply** [comment]
Agree

For programs the test/demo data is initialized after the initial sync so I usually have to delete the DB to get the latest data initialized... (I don't try to talk myself out of the main problem ;) )

**Pull request merged**